File: services/warehouse-query-py/credential.py
"""Tokens: the service's own identity, and the user's, on the user's behalf.

Two hops, both standard:

  1. **The service's identity** is a managed identity. It is discovered exactly
     as every Azure SDK discovers one — the `IDENTITY_ENDPOINT` /
     `IDENTITY_HEADER` App Service protocol, falling back to IMDS — so the same
     code runs in Container Apps, App Service and Functions with nothing set by
     us. It is used to read the executor's own secrets and (where the platform
     federates it) to prove the middle tier without a secret at all.

  2. **The user's identity** reaches the data plane by OAuth 2.0 On-Behalf-Of
     (RFC 7523 `jwt-bearer` + `requested_token_use=on_behalf_of`). The token the
     database sees carries the USER, so the database's own permissions apply.
     The middle tier authenticates itself with a federated client assertion
     (preferred, secretless) and falls back to a client secret **read from Key
     Vault with the managed identity** — never an environment variable.

Nothing here is emulator-aware. `DAS_ENTRA_TLS_INSECURE` is the family's
self-signed-certificate switch and is off in production.
"""

from __future__ import annotations

import dataclasses
import json
import os
import ssl
import threading
import time
import urllib.error
import urllib.parse
import urllib.request

EXCHANGE_AUDIENCE = "api://AzureADTokenExchange"


def _bool(name: str, default: str = "false") -> bool:
    return os.environ.get(name, default).strip().lower() in ("1", "true", "yes")


def _ssl_context() -> ssl.SSLContext:
    ctx = ssl.create_default_context()
    if _bool("DAS_ENTRA_TLS_INSECURE"):
        ctx.check_hostname = False
        ctx.verify_mode = ssl.CERT_NONE
    return ctx


_SSL = _ssl_context()


def _post_form(url: str, form: dict[str, str]) -> dict:
    data = urllib.parse.urlencode(form).encode()
    req = urllib.request.Request(
        url, data=data, method="POST", headers={"Content-Type": "application/x-www-form-urlencoded"}
    )
    try:
        with urllib.request.urlopen(req, context=_SSL, timeout=30) as r:
            return json.loads(r.read())
    except urllib.error.HTTPError as e:
        body = e.read().decode()
        raise TokenError(f"{e.code} {body[:400]}") from None


def _get_json(url: str, headers: dict[str, str]) -> dict:
    req = urllib.request.Request(url, headers=headers)
    try:
        with urllib.request.urlopen(req, context=_SSL, timeout=30) as r:
            return json.loads(r.read())
    except urllib.error.HTTPError as e:
        raise TokenError(f"{e.code} {e.read().decode()[:400]}") from None


class TokenError(Exception):
    pass


@dataclasses.dataclass(frozen=True)
class Settings:
    authority: str  # https://login.microsoftonline.com/<tenant>
    client_id: str  # the middle tier == the API app
    keyvault_url: str = ""
    secret_name: str = "das-executor-client-secret"
    identity_endpoint: str = ""
    identity_header: str = ""

    @staticmethod
    def from_env() -> Settings:
        issuer = os.environ["DAS_ENTRA_ISSUER"].rstrip("/")
        authority = issuer[: -len("/v2.0")] if issuer.endswith("/v2.0") else issuer
        return Settings(
            authority=authority,
            client_id=os.environ["DAS_MIDDLE_TIER_CLIENT_ID"],
            keyvault_url=os.environ.get("DAS_KEYVAULT_URL", "").rstrip("/"),
            secret_name=os.environ.get("DAS_EXECUTOR_SECRET_NAME", "das-executor-client-secret"),
            identity_endpoint=os.environ.get("IDENTITY_ENDPOINT", ""),
            identity_header=os.environ.get("IDENTITY_HEADER", ""),
        )


class Credential:
    """Caches tokens per (audience, user) until shortly before expiry."""

    def __init__(self, s: Settings):
        self.s = s
        self._lock = threading.Lock()
        self._cache: dict[tuple[str, str], tuple[float, str]] = {}
        self._secret: str | None = None
        self._secrets: dict[str, str] = {}

    # -- 1. the service's own identity ------------------------------------
    def managed_identity_token(self, resource: str) -> str:
        key = ("mi", resource)
        hit = self._cache.get(key)
        if hit and hit[0] - 60 > time.time():
            return hit[1]
        if not self.s.identity_endpoint:
            raise TokenError("no managed identity endpoint (IDENTITY_ENDPOINT unset)")
        url = f"{self.s.identity_endpoint}?resource={urllib.parse.quote(resource)}&api-version=2019-08-01"
        r = _get_json(url, {"X-IDENTITY-HEADER": self.s.identity_header})
        tok = r["access_token"]
        exp = float(r.get("expires_on") or (time.time() + 3600))
        with self._lock:
            self._cache[key] = (exp, tok)
        return tok

    # -- middle-tier credential -------------------------------------------
    def _client_assertion(self) -> str | None:
        try:
            return self.managed_identity_token(EXCHANGE_AUDIENCE)
        except TokenError:
            return None

    def _client_secret(self) -> str | None:
        """Read once from Key Vault, authenticating with the managed identity."""
        if self._secret is not None:
            return self._secret
        if not self.s.keyvault_url:
            return None
        try:
            tok = self.managed_identity_token("https://vault.azure.net")
            r = _get_json(
                f"{self.s.keyvault_url}/secrets/{self.s.secret_name}?api-version=7.5",
                {"Authorization": "Bearer " + tok},
            )
            self._secret = r["value"]
            return self._secret
        except (TokenError, KeyError):
            return None

    def secret(self, name: str) -> str | None:
        """Any secret from the vault, by name.

        HTTP sources need this: an API that does not federate with Entra
        cannot take an on-behalf-of token, so a stored credential is the only
        way to reach it — and a stored credential is exactly why such a source
        is `authz_tier=service` and says so in every audit line.
        """
        if not self.s.keyvault_url or not name:
            return None
        cached = self._secrets.get(name)
        if cached is not None:
            return cached
        try:
            tok = self.managed_identity_token("https://vault.azure.net")
            r = _get_json(
                f"{self.s.keyvault_url}/secrets/{name}?api-version=7.5",
                {"Authorization": "Bearer " + tok},
            )
        except (TokenError, KeyError):
            return None
        value = r.get("value")
        if value:
            self._secrets[name] = value
        return value

    # -- 2. the user's identity, on their behalf ---------------------------
    def on_behalf_of(self, user_assertion: str, scope: str, cache_key: str = "") -> str:
        key = (scope, cache_key or user_assertion[-32:])
        hit = self._cache.get(key)
        if hit and hit[0] - 60 > time.time():
            return hit[1]
        form = {
            "grant_type": "urn:ietf:params:oauth:grant-type:jwt-bearer",
            "client_id": self.s.client_id,
            "assertion": user_assertion,
            "scope": scope,
            "requested_token_use": "on_behalf_of",
        }
        url = f"{self.s.authority}/oauth2/v2.0/token"
        assertion = self._client_assertion()
        errors = []
        if assertion:
            try:
                r = _post_form(
                    url,
                    {
                        **form,
                        "client_assertion_type": "urn:ietf:params:oauth:client-assertion-type:jwt-bearer",
                        "client_assertion": assertion,
                    },
                )
                return self._store(key, r)
            except TokenError as e:
                errors.append(f"federated credential: {e}")
        secret = self._client_secret()
        if secret:
            try:
                r = _post_form(url, {**form, "client_secret": secret})
                return self._store(key, r)
            except TokenError as e:
                errors.append(f"client secret: {e}")
        raise TokenError("on-behalf-of exchange failed — " + "; ".join(errors or ["no credential"]))

    def _store(self, key, r: dict) -> str:
        tok = r["access_token"]
        with self._lock:
            self._cache[key] = (time.time() + int(r.get("expires_in", 3600)), tok)
        return tok

File: services/warehouse-query-go/credential.go
// Tokens: the service's own identity, and the user's, on the user's behalf.
// The Go half of services/warehouse-query-py/credential.py — same two hops,
// same standard protocols, same order of preference (federated first, a Key
// Vault secret second, never an environment variable).
package main

import (
	"context"
	"crypto/tls"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"strconv"
	"strings"
	"sync"
	"time"
)

const exchangeAudience = "api://AzureADTokenExchange"

var insecureTLS = strings.EqualFold(os.Getenv("DAS_ENTRA_TLS_INSECURE"), "true")

func httpClient() *http.Client {
	transport := &http.Transport{}
	if insecureTLS {
		transport.TLSClientConfig = &tls.Config{InsecureSkipVerify: true} // #nosec G402 — dev certificates only
	}
	return &http.Client{Timeout: 60 * time.Second, Transport: transport}
}

var client = httpClient()

type Credential struct {
	authority   string
	clientID    string
	keyVault    string
	secretName  string
	miEndpoint  string
	miHeader    string
	mu          sync.Mutex
	tokens      map[string]cachedToken
	secret      string
	secretKnown bool
}

type cachedToken struct {
	value     string
	expiresAt time.Time
}

func NewCredential() *Credential {
	issuer := strings.TrimSuffix(os.Getenv("DAS_ENTRA_ISSUER"), "/")
	authority := strings.TrimSuffix(issuer, "/v2.0")
	secretName := os.Getenv("DAS_EXECUTOR_SECRET_NAME")
	if secretName == "" {
		// The NAME of a secret in Key Vault, not a secret. gosec cannot tell
		// the difference, and a name is exactly what belongs in code.
		secretName = "das-executor-client-secret" //nolint:gosec // G101: vault key name
	}
	return &Credential{
		authority:  authority,
		clientID:   os.Getenv("DAS_MIDDLE_TIER_CLIENT_ID"),
		keyVault:   strings.TrimSuffix(os.Getenv("DAS_KEYVAULT_URL"), "/"),
		secretName: secretName,
		miEndpoint: os.Getenv("IDENTITY_ENDPOINT"),
		miHeader:   os.Getenv("IDENTITY_HEADER"),
		tokens:     map[string]cachedToken{},
	}
}

// How long any single identity call may take. Short on purpose: a token
// endpoint that has stopped answering should surface as a failed request, not
// as a request that never returns.
const tokenCallTimeout = 30 * time.Second

func (c *Credential) cached(key string) (string, bool) {
	c.mu.Lock()
	defer c.mu.Unlock()
	if hit, ok := c.tokens[key]; ok && time.Until(hit.expiresAt) > time.Minute {
		return hit.value, true
	}
	return "", false
}

func (c *Credential) store(key, value string, expires time.Time) {
	c.mu.Lock()
	c.tokens[key] = cachedToken{value: value, expiresAt: expires}
	c.mu.Unlock()
}

// ManagedIdentityToken uses the App Service protocol — the same two variables
// the platform sets in Azure, so this code is unchanged in production.
func (c *Credential) ManagedIdentityToken(resource string) (string, error) {
	key := "mi:" + resource
	if token, ok := c.cached(key); ok {
		return token, nil
	}
	if c.miEndpoint == "" {
		return "", fmt.Errorf("no managed identity endpoint (IDENTITY_ENDPOINT unset)")
	}
	target := c.miEndpoint + "?resource=" + url.QueryEscape(resource) + "&api-version=2019-08-01"
	// Every outbound call carries a deadline it can be cancelled by. A client
	// timeout alone stops a hung response; it does not stop a caller that has
	// already given up from holding the connection.
	ctx, cancel := context.WithTimeout(context.Background(), tokenCallTimeout)
	defer cancel()
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, target, nil)
	if err != nil {
		return "", err
	}
	req.Header.Set("X-IDENTITY-HEADER", c.miHeader)
	var payload struct {
		AccessToken string `json:"access_token"`
		ExpiresOn   string `json:"expires_on"`
	}
	if err := doJSON(req, &payload); err != nil {
		return "", err
	}
	expires := time.Now().Add(time.Hour)
	if seconds, err := strconv.ParseInt(payload.ExpiresOn, 10, 64); err == nil {
		expires = time.Unix(seconds, 0)
	}
	c.store(key, payload.AccessToken, expires)
	return payload.AccessToken, nil
}

func (c *Credential) clientSecret() string {
	c.mu.Lock()
	if c.secretKnown {
		defer c.mu.Unlock()
		return c.secret
	}
	c.mu.Unlock()
	value := ""
	if c.keyVault != "" {
		if token, err := c.ManagedIdentityToken("https://vault.azure.net"); err == nil {
			ctx, cancel := context.WithTimeout(context.Background(), tokenCallTimeout)
			defer cancel()
			req, _ := http.NewRequestWithContext(ctx, http.MethodGet,
				c.keyVault+"/secrets/"+c.secretName+"?api-version=7.5", nil)
			req.Header.Set("Authorization", "Bearer "+token)
			var payload struct {
				Value string `json:"value"`
			}
			if err := doJSON(req, &payload); err == nil {
				value = payload.Value
			}
		}
	}
	c.mu.Lock()
	c.secret, c.secretKnown = value, true
	c.mu.Unlock()
	return value
}

// OnBehalfOf exchanges the USER's token for a data-plane token that still
// carries the user, so the source applies that user's own permissions.
func (c *Credential) OnBehalfOf(userAssertion, scope, cacheKey string) (string, error) {
	key := scope + "|" + cacheKey
	if token, ok := c.cached(key); ok {
		return token, nil
	}
	form := url.Values{
		"grant_type":          {"urn:ietf:params:oauth:grant-type:jwt-bearer"},
		"client_id":           {c.clientID},
		"assertion":           {userAssertion},
		"scope":               {scope},
		"requested_token_use": {"on_behalf_of"},
	}
	var problems []string

	if assertion, err := c.ManagedIdentityToken(exchangeAudience); err == nil {
		federated := cloneValues(form)
		federated.Set("client_assertion_type",
			"urn:ietf:params:oauth:client-assertion-type:jwt-bearer")
		federated.Set("client_assertion", assertion)
		if token, expires, err := c.tokenRequest(federated); err == nil {
			c.store(key, token, expires)
			return token, nil
		} else {
			problems = append(problems, "federated credential: "+err.Error())
		}
	}
	if secret := c.clientSecret(); secret != "" {
		withSecret := cloneValues(form)
		withSecret.Set("client_secret", secret)
		if token, expires, err := c.tokenRequest(withSecret); err == nil {
			c.store(key, token, expires)
			return token, nil
		} else {
			problems = append(problems, "client secret: "+err.Error())
		}
	}
	if len(problems) == 0 {
		problems = append(problems, "no credential")
	}
	return "", fmt.Errorf("on-behalf-of exchange failed — %s", strings.Join(problems, "; "))
}

func cloneValues(in url.Values) url.Values {
	out := url.Values{}
	for k, v := range in {
		out[k] = append([]string(nil), v...)
	}
	return out
}

func (c *Credential) tokenRequest(form url.Values) (string, time.Time, error) {
	ctx, cancel := context.WithTimeout(context.Background(), tokenCallTimeout)
	defer cancel()
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, c.authority+"/oauth2/v2.0/token",
		strings.NewReader(form.Encode()))
	if err != nil {
		return "", time.Time{}, err
	}
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	var payload struct {
		AccessToken string `json:"access_token"`
		ExpiresIn   int    `json:"expires_in"`
	}
	if err := doJSON(req, &payload); err != nil {
		return "", time.Time{}, err
	}
	seconds := payload.ExpiresIn
	if seconds == 0 {
		seconds = 3600
	}
	return payload.AccessToken, time.Now().Add(time.Duration(seconds) * time.Second), nil
}

func doJSON(req *http.Request, out any) error {
	resp, err := client.Do(req)
	if err != nil {
		return err
	}
	defer func() { _ = resp.Body.Close() }()
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return err
	}
	if resp.StatusCode >= 400 {
		return fmt.Errorf("%d %s", resp.StatusCode, truncate(string(body), 300))
	}
	return json.Unmarshal(body, out)
}

func getJSON(target, token string, out any) error {
	ctx, cancel := context.WithTimeout(context.Background(), tokenCallTimeout)
	defer cancel()
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, target, nil)
	if err != nil {
		return err
	}
	req.Header.Set("Authorization", "Bearer "+token)
	return doJSON(req, out)
}

func truncate(s string, n int) string {
	if len(s) <= n {
		return s
	}
	return s[:n]
}
